# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxUtils.cpp
# Commit: 71c122ac0ca7
# Full Hash: 71c122ac0ca73391866b1ef19f4f82bc2d28568b
# Author: David Parks <davidp99@gmail.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Initial out-of-process WebGL implementation. r=mccr8,handyman
#   
#   Splits WebGLContext into ClientWebGLContext and HostWebGLContext.  The Client enables the JS-control of a WebGL context in a content procecss while the Host executes the WebGL graphics operations (via a WebGLContext that maintains much of the existing code) in the compositor process.  At this point, the cross-process behavior is disabled -- this series of patches is an incremental step toward that final goal.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D54018
# ==============================================================================

diff -r 5cbf5bb11c58 -r 71c122ac0ca7 gfx/thebes/gfxUtils.cpp
--- a/gfx/thebes/gfxUtils.cpp	Wed Jan 08 21:52:03 2020 +0000
+++ b/gfx/thebes/gfxUtils.cpp	Wed Jan 08 22:19:14 2020 +0000
@@ -17,6 +17,7 @@
 #include "mozilla/dom/ImageEncoder.h"
 #include "mozilla/dom/WorkerPrivate.h"
 #include "mozilla/dom/WorkerRunnable.h"
+#include "mozilla/ipc/CrossProcessSemaphore.h"
 #include "mozilla/gfx/2D.h"
 #include "mozilla/gfx/DataSurfaceHelpers.h"
 #include "mozilla/gfx/Logging.h"
@@ -27,6 +28,7 @@
 #include "mozilla/image/nsICOEncoder.h"
 #include "mozilla/image/nsJPEGEncoder.h"
 #include "mozilla/image/nsPNGEncoder.h"
+#include "mozilla/layers/SynchronousTask.h"
 #include "mozilla/Maybe.h"
 #include "mozilla/RefPtr.h"
 #include "mozilla/StaticPrefs_gfx.h"
@@ -1356,11 +1358,13 @@
       format, encoder, aEncoderOptions, outStream);
 }
 
-class GetFeatureStatusRunnable final : public dom::WorkerMainThreadRunnable {
+class GetFeatureStatusWorkerRunnable final
+    : public dom::WorkerMainThreadRunnable {
  public:
-  GetFeatureStatusRunnable(dom::WorkerPrivate* workerPrivate,
-                           const nsCOMPtr<nsIGfxInfo>& gfxInfo, int32_t feature,
-                           nsACString& failureId, int32_t* status)
+  GetFeatureStatusWorkerRunnable(dom::WorkerPrivate* workerPrivate,
+                                 const nsCOMPtr<nsIGfxInfo>& gfxInfo,
+                                 int32_t feature, nsACString& failureId,
+                                 int32_t* status)
       : WorkerMainThreadRunnable(workerPrivate,
                                  NS_LITERAL_CSTRING("GFX :: GetFeatureStatus")),
         mGfxInfo(gfxInfo),
@@ -1379,7 +1383,7 @@
   nsresult GetNSResult() const { return mNSResult; }
 
  protected:
-  ~GetFeatureStatusRunnable() {}
+  ~GetFeatureStatusWorkerRunnable() {}
 
  private:
   nsCOMPtr<nsIGfxInfo> mGfxInfo;
@@ -1393,11 +1397,23 @@
 nsresult gfxUtils::ThreadSafeGetFeatureStatus(
     const nsCOMPtr<nsIGfxInfo>& gfxInfo, int32_t feature, nsACString& failureId,
     int32_t* status) {
-  if (!NS_IsMainThread()) {
+  if (NS_IsMainThread()) {
+    return gfxInfo->GetFeatureStatus(feature, failureId, status);
+  }
+
+  // In a content process, we must call this on the main thread.
+  // In a composition process (parent or GPU), this needs to be called on the
+  // compositor thread.
+  bool isCompositionProcess = XRE_IsGPUProcess() || XRE_IsParentProcess();
+  MOZ_ASSERT(!isCompositionProcess || NS_IsInCompositorThread());
+
+  // Content-process non-main-thread case:
+  if (!isCompositionProcess) {
     dom::WorkerPrivate* workerPrivate = dom::GetCurrentThreadWorkerPrivate();
 
-    RefPtr<GetFeatureStatusRunnable> runnable = new GetFeatureStatusRunnable(
-        workerPrivate, gfxInfo, feature, failureId, status);
+    RefPtr<GetFeatureStatusWorkerRunnable> runnable =
+        new GetFeatureStatusWorkerRunnable(workerPrivate, gfxInfo, feature,
+                                           failureId, status);
 
     ErrorResult rv;
     runnable->Dispatch(dom::WorkerStatus::Canceling, rv);
@@ -1407,11 +1423,17 @@
       // exception.  Ah, well.
       return rv.StealNSResult();
     }
-
     return runnable->GetNSResult();
   }
 
-  return gfxInfo->GetFeatureStatus(feature, failureId, status);
+  nsresult rv;
+  SynchronousTask task("GetFeatureStatusSync");
+  NS_DispatchToMainThread(NS_NewRunnableFunction("GetFeatureStatusMain", [&]() {
+    AutoCompleteTask complete(&task);
+    rv = gfxInfo->GetFeatureStatus(feature, failureId, status);
+  }));
+  task.Wait();
+  return rv;
 }
 
 #define GFX_SHADER_CHECK_BUILD_VERSION_PREF "gfx-shader-check.build-version"